cpp/src/arrow/compute/expression.h

```diff
@@ -159,14 +171,15 @@ Expression field_ref(FieldRef ref);
 
 ARROW_EXPORT
 Expression call(std::string function, std::vector<Expression> arguments,
-                std::shared_ptr<FunctionOptions> options = NULLPTR);
+                std::shared_ptr<FunctionOptions> options = NULLPTR,
+                bool is_special_form = false);
 
 template <typename Options, typename = typename std::enable_if<
                                 std::is_base_of<FunctionOptions, Options>::value>::type>
-Expression call(std::string function, std::vector<Expression> arguments,
-                Options options) {
+Expression call(std::string function, std::vector<Expression> arguments, Options options,
+                bool is_special_form = false) {
```

Comment:
Special forms are not "called", they are evaluated. They are more general expressions than calls. `if (true) f(x)` is equivalent to a call to `f` with `x` that you discover while trying to evaluate the conditional.